### PR TITLE
NEW: Extension points for FilesystemPublisher.

### PR DESCRIPTION
**Pull request by mfendeksilverstripe**
# NEW: Extension points for FilesystemPublisher.

* Extension points added to `generatePageResponse()` to allow better extensibility

### CODE REVIEW

**Comment by dhensby** [comment]
The idea behind the publisher classes is that developers can easily implement their own and add their own logic easily. There shouldn't really be a need for extension hooks with the current implementation.

What is the scenario where this is needed?

**Comment by mfendeksilverstripe** [comment]
@dhensby My team found useful to have the ability to inject some custom behaviour via these extension points. For example we have in-memory cache which needs to be flushed after each call of `generatePageResponse()`. It's not something that warrants the implementation of a new file system publisher as we're not altering any core behaviour but instead a simple extension can accomodate this requirement.

Also, we found useful to reset some core states such as controller stack something like this:

```php
try {
    $controllers = [];

    // pop all current application's controllers to simulate cold boot
    while (Controller::has_curr()) {
        $controller = Controller::curr();
        $controllers[]= $controller;
        $controller->popCurrent();
    }

    return parent::generatePageResponse($url);
} finally {
    // flush all in-memory cached data

    foreach (array_reverse($controllers) as $controller) {
        // reapply previous controllers
        $controller->pushCurrent();
    }
}
```

File: src/Publisher/FilesystemPublisher.php [hunk 1]
```diff
@@ -68,7 +68,7 @@ public function purgeURL($url)
     {
         if (!$url) {
             user_error('Bad url:' . var_export($url, true), E_USER_WARNING);
-            return;
+            return [];
```
Comment (ChloeSartorelli) [inline src/Publisher/FilesystemPublisher.php]: Why is this change here? It seems out of scope for adding extension points. If there is a valid reason to change what is returned here, please create a separate PR for this purpose which can be reviewed and merged on its own merits.
```suggestion
```

File: src/Publisher/FilesystemPublisher.php [hunk 2]
```diff
@@ -101,10 +101,14 @@ public function publishURL($url, $forcePublish = false)
...
```
Comment (ChloeSartorelli) [inline src/Publisher/FilesystemPublisher.php]:
```suggestion
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Your use case seems sensible to me. Please retarget this PR to the `6` branch and remove the unrelated changes.

**Comment by mfendeksilverstripe** [comment]
I've pushed up changes as per your request @GuySartorelli , please review.

**Comment by GuySartorelli** [comment]
You swapped the target branch _after_ resetting commits, which means the new CI isn't able to run. Can you please force push so that the CI runs correctly?

**Comment by mfendeksilverstripe** [comment]
Done @GuySartorelli

**Review by GuySartorelli — APPROVED** [review]
Changes look good to me, will merge when CI goes green.

**Pull request merged**